CMakeLists.txt

```diff
@@ -644,74 +557,48 @@ if(USE_GCC OR USE_CLANG OR USE_INTELCC OR USE_QCC)
 
-  check_c_compiler_flag(-Wshadow HAVE_GCC_WSHADOW)
-  if(HAVE_GCC_WSHADOW)
-    target_compile_options(sdl-global-options INTERFACE "-Wshadow")
-  endif()
-
-  check_c_compiler_flag(-Wunused-local-typedefs HAVE_GCC_WUNUSED_LOCAL_TYPEDEFS)
-  if(HAVE_GCC_WUNUSED_LOCAL_TYPEDEFS)
-    target_compile_options(sdl-global-options INTERFACE "-Wno-unused-local-typedefs")
-  endif()
-
   if(APPLE)
-    cmake_push_check_state()
     # FIXME: don't use deprecated declarations
-    check_c_compiler_flag(-Wno-error=deprecated-declarations HAVE_WNO_ERROR_DEPRECATED_DECLARATIONS)
-    if(HAVE_WNO_ERROR_DEPRECATED_DECLARATIONS)
-      target_compile_options(sdl-build-options INTERFACE "-Wno-error=deprecated-declarations")
-    endif()
-    cmake_pop_check_state()
-
-    if(CMAKE_VERSION VERSION_LESS "3.6")
-      target_link_libraries(sdl-shared-build-options INTERFACE "-Wl,-compatibility_version,${SDL_DYLIB_COMPAT_VERSION}")
-      target_link_libraries(sdl-shared-build-options INTERFACE "-Wl,-current_version,${SDL_DYLIB_CURRENT_VERSION}")
-    endif()
-  elseif(NOT OPENBSD)
-    cmake_push_check_state()
-    set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -Wl,--no-undefined")
-    check_c_compiler_flag("" HAVE_NO_UNDEFINED)
-    cmake_pop_check_state()
-    if(HAVE_NO_UNDEFINED AND NOT (USE_CLANG AND WINDOWS))
-      target_link_libraries(sdl-shared-build-options INTERFACE "-Wl,--no-undefined")
+    check_c_compiler_flag(-Wno-error=deprecated-declarations COMPILER_SUPPORTS_WNO_ERROR_DEPRECATED_DECLARATIONS)
+    if(COMPILER_SUPPORTS_WNO_ERROR_DEPRECATED_DECLARATIONS)
+      sdl_compile_options(PRIVATE "-Wno-error=deprecated-declarations")
     endif()
   endif()
 
-  if(MINGW)
-    # See if GCC's -gdwarf-4 is supported
-    # See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101377 for why this is needed on Windows
-    cmake_push_check_state()
-    check_c_compiler_flag("-gdwarf-4" HAVE_GDWARF_4)
-    if(HAVE_GDWARF_4)
-      target_compile_options(sdl-global-options INTERFACE "-gdwarf-4")
+  if(APPLE)
+    check_linker_flag(C "-Wl,-undefined,error" COMPILER_SUPPORTS_WL_UNDEFINED_ERROR)
+    if(COMPILER_SUPPORTS_WL_UNDEFINED_ERROR)
+      sdl_shared_link_options("-Wl,-undefined,error")
     endif()
-    cmake_pop_check_state()
-  endif()
-
-  # Force color diagnostics when one of these conditions are met
-  if(DEFINED ENV{CI} OR DEFINED ENV{USE_CCACHE} OR CMAKE_GENERATOR MATCHES Ninja)
-    if(EMSCRIPTEN OR (USE_GCC AND NOT CMAKE_C_COMPILER_VERSION VERSION_LESS 4.9))
-      target_compile_options(sdl-global-options INTERFACE -fdiagnostics-color=always)
-    elseif(USE_CLANG AND NOT CMAKE_C_COMPILER_ID MATCHES AppleClang)
-      target_compile_options(sdl-global-options INTERFACE -fcolor-diagnostics)
+  else()
+    cmake_push_check_state()
+    check_linker_flag(C "-Wl,--no-undefined" COMPILER_SUPPORTS_WL_NO_UNDEFINED)
+    #FIXME: originally this if had an additional "AND NOT (USE_CLANG AND WINDOWS)"
+    if(COMPILER_SUPPORTS_WL_NO_UNDEFINED)
```

Comment:
Are we sure openbsd behaves after this?

Reply:
No, because it's not tested on ci.
The `if` is only taken if the linker accepts `-Wl,--no-undefined`.

At what potential issue are you hinting?

Reply:
Autotools (which cmake inherited the original check) specifically rejects openbsd noting it among the known broken, i.e. the linker check will probably pass even if it is actually broken.

P.S.: The cmake variable name should probably be changed from `COMPILER_SUPPORTS_*` to `LINKER_SUPPORTS__*`

Reply:
Original commit noting the issue: https://github.com/SDL-mirror/SDL/commit/05b7b71074b5e66e330bf063e01097e2e182b982

Reply:
I made an exception for openbsd.
I found similar issues on the mesa and meson issue trackers.